File: src/predict/events.c
#include "colyseus/predict/events.h"
#include "uthash.h"

#include <math.h>
#include <stdlib.h>
#include <string.h>

/* colyseus_reconciler internals we reach through the step ctx backing */
colyseus_input_handle_t* colyseus_reconciler_input_(const void* reconciler);

typedef struct {
    char* key;                      /* owned */
    void* payload;
    int seq;                        /* birth seq; -1 = UI-born */
    colyseus_input_handle_t* acked; /* sim-born: the emitting handle (watermark) */
    double at;                      /* birth instant (serverNow axis) */
    UT_hash_handle hh;
} event_entry_t;

struct colyseus_event_channel {
    colyseus_event_channel_options_t opts;
    colyseus_room_clock_t* clock;
    event_entry_t* entries;
    double cooldown_until;
};

static double channel_now(const colyseus_event_channel_t* channel) {
    return channel->clock
        ? colyseus_room_clock_server_now(channel->clock)
        : colyseus_room_clock_now(NULL);
}

colyseus_event_channel_t* colyseus_event_channel_create(
    const colyseus_event_channel_options_t* options,
    colyseus_room_clock_t* clock) {
    colyseus_event_channel_t* channel = calloc(1, sizeof(colyseus_event_channel_t));
    if (options) channel->opts = *options;
    if (channel->opts.grace_ticks <= 0) channel->opts.grace_ticks = 10;
    channel->clock = clock;
    channel->cooldown_until = -INFINITY;
    return channel;
}

static void drop_entry(colyseus_event_channel_t* channel, event_entry_t* entry, bool free_payload) {
    HASH_DEL(channel->entries, entry);
    if (free_payload && channel->opts.payload_free) channel->opts.payload_free(entry->payload);
    free(entry->key);
    free(entry);
}

void colyseus_event_channel_free(colyseus_event_channel_t* channel) {
    if (!channel) return;
    colyseus_event_channel_clear(channel);
    free(channel);
}

static bool channel_add(colyseus_event_channel_t* channel, const char* key,
    void* payload, int seq, colyseus_input_handle_t* acked) {
    event_entry_t* existing = NULL;
    HASH_FIND_STR(channel->entries, key, existing);
    if (existing) return false;   /* already pending — re-derivations don't re-fire */

    double t = channel_now(channel);
    if (channel->opts.cooldown_ms > 0) {
        if (t < channel->cooldown_until) return false;
        channel->cooldown_until = t + channel->opts.cooldown_ms;
    }

    event_entry_t* entry = calloc(1, sizeof(event_entry_t));
    entry->key = strdup(key);
    entry->payload = payload;
    entry->seq = seq;
    entry->acked = acked;
    entry->at = t;
    HASH_ADD_KEYPTR(hh, channel->entries, entry->key, strlen(entry->key), entry);

    if (channel->opts.on_predict) channel->opts.on_predict(payload, channel->opts.userdata);
    return true;
}

bool colyseus_event_channel_predict(colyseus_event_channel_t* channel,
    const char* key, void* payload) {
    if (!channel || !key) return false;
    return channel_add(channel, key, payload, -1, NULL);
}

void colyseus_step_predict(const colyseus_step_ctx_t* ctx,
    colyseus_event_channel_t* channel, const char* key, void* payload) {
    if (!ctx || !channel || !key) return;
    if (ctx->is_replay) return;   /* live-only by construction */
    colyseus_input_handle_t* handle = ctx->_memo_backing
        ? colyseus_reconciler_input_(ctx->_memo_backing)
        : NULL;
    channel_add(channel, key, payload, ctx->tick, handle);
}

bool colyseus_event_channel_has(const colyseus_event_channel_t* channel, const char* key) {
    if (!channel) return false;
    if (!key) return channel->entries != NULL;
    event_entry_t* entry = NULL;
    HASH_FIND_STR(((colyseus_event_channel_t*)channel)->entries, key, entry);
    return entry != NULL;
}

int colyseus_event_channel_pending_count(const colyseus_event_channel_t* channel) {
    return channel ? (int)HASH_COUNT(channel->entries) : 0;
}

int colyseus_event_channel_confirm(colyseus_event_channel_t* channel, const char* key) {
    if (!channel) return 0;
    int settled = 0;
    event_entry_t *entry, *tmp;
    HASH_ITER(hh, channel->entries, entry, tmp) {
        if (key != NULL && strcmp(entry->key, key) != 0) continue;
        void* payload = entry->payload;
        drop_entry(channel, entry, false);   /* removed BEFORE the callback */
        if (channel->opts.on_confirm) channel->opts.on_confirm(payload, channel->opts.userdata);
        if (channel->opts.payload_free) channel->opts.payload_free(payload);
        settled++;
    }
    if (settled == 0 && channel->opts.on_unpredicted) {
        channel->opts.on_unpredicted(key, channel->opts.userdata);
    }
    return settled;
}

static void reject_entry(colyseus_event_channel_t* channel, event_entry_t* entry) {
    void* payload = entry->payload;
    drop_entry(channel, entry, false);
    if (channel->opts.on_reject) channel->opts.on_reject(payload, channel->opts.userdata);
    if (channel->opts.payload_free) channel->opts.payload_free(payload);
}

int colyseus_event_channel_reject(colyseus_event_channel_t* channel, const char* key) {
    if (!channel) return 0;
    int rejected = 0;
    event_entry_t *entry, *tmp;
    HASH_ITER(hh, channel->entries, entry, tmp) {
        if (key != NULL && strcmp(entry->key, key) != 0) continue;
        reject_entry(channel, entry);
        rejected++;
    }
    return rejected;
}

void colyseus_event_channel_clear(colyseus_event_channel_t* channel) {
    if (!channel) return;
    event_entry_t *entry, *tmp;
    HASH_ITER(hh, channel->entries, entry, tmp) {
        drop_entry(channel, entry, true);
    }
}

void colyseus_event_channel_prune(colyseus_event_channel_t* channel) {
    if (!channel || channel->entries == NULL) return;

    /* sim-born grace auto-reject: the server processed grace ticks past the
     * prediction without confirming — the event didn't happen */
    event_entry_t *entry, *tmp;
    HASH_ITER(hh, channel->entries, entry, tmp) {
        if (entry->acked == NULL) continue;
        if (colyseus_input_handle_last_processed(entry->acked) >= entry->seq + channel->opts.grace_ticks) {
            reject_entry(channel, entry);
        }
    }

    /* wall-clock TTL — UI-born entries only (sim-born settle by progress) */
    double rtt = channel->clock ? colyseus_room_clock_smoothed_rtt(channel->clock) : 0;
    double ttl = channel->opts.ttl_ms > 0 ? channel->opts.ttl_ms : (rtt * 2 > 600 ? rtt * 2 : 600);
    double now = channel_now(channel);
    HASH_ITER(hh, channel->entries, entry, tmp) {
        if (entry->acked != NULL) continue;   /* sim-born: progress-settled */
        if (now - entry->at > ttl) reject_entry(channel, entry);
    }
}

File: platforms/flutter/colyseus_flutter/lib/src/events.dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';

import 'bindings/colyseus_core.dart';
import 'colyseus.dart';
import 'interned_names.dart';
import 'payload_registry.dart';

/// A channel of optimistic events of one logical kind.
///
/// Some things a game does aren't positions — scoring a goal, firing a weapon,
/// picking something up. Prediction still applies: show it immediately, then
/// let the server agree or overrule. This channel tracks predictions the
/// client has made but the server hasn't settled.
///
/// Predict from the simulation with [StepContext.predict], which is
/// replay-safe (a rollback replay won't re-fire it), or from UI code with
/// [predict]. Settle with [confirm] when the authoritative signal arrives:
///
/// ```dart
/// final goals = predict.defineEvent(
///   onPredict: (payload) => showGoalBanner(),
///   onReject: (payload) => retractGoalBanner(),
/// );
///
/// // inside the reconciler step, when the predicted position crosses the line
/// ctx.predict(goals, 'goal:${ctx.tick}');
///
/// room.onMessage('goal').listen((_) => goals.confirm());
/// ```
///
/// Unsettled predictions don't linger: sim-born ones auto-reject once the
/// server has processed past them, UI-born ones expire on a wall clock.
class EventChannel {
  final Pointer<colyseus_event_channel> _handle;
  final List<NativeCallable> _callables;
  bool _disposed = false;

  EventChannel._(this._handle, this._callables);

  /// Builds a channel. Prefer [Predict.defineEvent], which also drives it.
  static EventChannel create({
    required Pointer<colyseus_room_clock> clock,
    void Function(Object? payload)? onPredict,
    void Function(Object? payload)? onConfirm,
    void Function(Object? payload)? onReject,
    void Function(String key)? onUnpredicted,
    int graceTicks = 0,
    double ttlMs = 0,
    double cooldownMs = 0,
  }) {
    final callables = <NativeCallable>[];

    NativeCallable<Void Function(Pointer<Void>, Pointer<Void>)>? wrap(
      void Function(Object?)? handler, {
      required bool releases,
    }) {
      if (handler == null && !releases) return null;
      final callable =
          NativeCallable<Void Function(Pointer<Void>, Pointer<Void>)>
              .isolateLocal((Pointer<Void> payload, Pointer<Void> _) {
        try {
          handler?.call(payloadFor(payload));
        } catch (_) {
          // Never unwind into C.
        }
      });
      callables.add(callable);
      return callable;
    }

    final predictCb = wrap(onPredict, releases: false);
    final confirmCb = wrap(onConfirm, releases: false);
    final rejectCb = wrap(onReject, releases: false);

    NativeCallable<Void Function(Pointer<Char>, Pointer<Void>)>? unpredictedCb;
    if (onUnpredicted != null) {
      unpredictedCb =
          NativeCallable<Void Function(Pointer<Char>, Pointer<Void>)>
              .isolateLocal((Pointer<Char> key, Pointer<Void> _) {
        try {
          onUnpredicted(key == nullptr ? '' : key.cast<Utf8>().toDartString());
        } catch (_) {}
      });
      callables.add(unpredictedCb);
    }

    // Runs after a settle callback and on silent drops — the one place a
    // payload's registry slot can be released without leaking or double-free.
    final freeCb = NativeCallable<Void Function(Pointer<Void>)>.isolateLocal(
      (Pointer<Void> payload) => releasePayload(payload),
    );
    callables.add(freeCb);

    final handle = using((arena) {
      final options = arena<colyseus_event_channel_options_t>();
      if (predictCb != null) options.ref.on_predict = predictCb.nativeFunction;
      if (confirmCb != null) options.ref.on_confirm = confirmCb.nativeFunction;
      if (rejectCb != null) options.ref.on_reject = rejectCb.nativeFunction;
      if (unpredictedCb != null) {
        options.ref.on_unpredicted = unpredictedCb.nativeFunction;
      }
      options.ref.payload_free = freeCb.nativeFunction;
      options.ref.grace_ticks = graceTicks;
      options.ref.ttl_ms = ttlMs;
      options.ref.cooldown_ms = cooldownMs;
      return core.colyseus_event_channel_create(options, clock);
    });

    if (handle == nullptr) {
      for (final callable in callables) {
        callable.close();
      }
      throw StateError('Could not create an event channel');
    }

    return EventChannel._(handle, callables);
  }

  /// The native handle.
  Pointer<colyseus_event_channel> get handle => _handle;

  /// Predicts an event from outside the simulation, for UI-driven optimism.
  ///
  /// Returns false when the prediction was dropped — [key] is already pending,
  /// or a cooldown is in effect. Settles on a wall clock, unlike the sim-born
  /// form, which settles on server progress.
  bool predict(String key, [Object? payload]) {
    return core.colyseus_event_channel_predict(
      _handle,
      internedName(key),
      retainPayload(payload),
    );
  }

  /// Whether [key] (or anything, when null) is awaiting settlement.
  bool has([String? key]) => core.colyseus_event_channel_has(
        _handle,
        key == null ? nullptr : internedName(key),
      );

  /// How many predictions are awaiting settlement.
  int get pendingCount => core.colyseus_event_channel_pending_count(_handle);

  /// Settles [key] (or every pending prediction) as correct.
  ///
  /// Returns how many were settled. Zero means the server signalled something
  /// this client never predicted, which fires `onUnpredicted`.
  int confirm([String? key]) => core.colyseus_event_channel_confirm(
        _handle,
        key == null ? nullptr : internedName(key),
      );

  /// Settles [key] (or every pending prediction) as wrong, firing `onReject`.
  int reject([String? key]) => core.colyseus_event_channel_reject(
        _handle,
        key == null ? nullptr : internedName(key),
      );

  /// Drops every pending prediction without firing confirm or reject.
  void clear() => core.colyseus_event_channel_clear(_handle);

  /// Releases the channel.
  void dispose() {
    if (_disposed) return;
    _disposed = true;

    core.colyseus_event_channel_free(_handle);
    for (final callable in _callables) {
      callable.close();
    }
    _callables.clear();
  }
}
